src/hostapi/pulseaudio/pa_linux_pulseaudio.c

```diff
@@ -0,0 +1,1313 @@
+     hostOutputSampleFormat;
+
+    /* validate platform specific flags */
+    if( (streamFlags & paPlatformSpecificFlags) != 0 )
+    {
+        return paInvalidFlag;   /* unexpected platform specific flag */
+    }
+
+
+    pa_threaded_mainloop_lock(l_ptrPulseAudioHostApi->mainloop);
+    stream =
+        (PaPulseAudio_Stream *) PaUtil_AllocateMemory( sizeof( PaPulseAudio_Stream ) );
+
+    if( !stream )
+    {
+        result = paInsufficientMemory;
+        goto error;
+    }
+
+    stream->isActive = 0;
+    stream->isStopped = 1;
+
+    stream->inStream = NULL;
+    stream->outStream = NULL;
+    stream->outBuffer = NULL;
+    stream->inBuffer = NULL;
+    memset( &stream->outputRing,
+            0x00,
+            sizeof(PaUtilRingBuffer) );
+    memset( &stream->inputRing,
+            0x00,
+            sizeof(PaUtilRingBuffer) );
+
+
+    if( inputParameters )
+    {
+        inputChannelCount = inputParameters->channelCount;
+        inputSampleFormat = inputParameters->sampleFormat;
+
+        /* unless alternate device specification is supported, reject the use of
+         * paUseHostApiSpecificDeviceSpecification
+         */
+        if( inputParameters->device == paUseHostApiSpecificDeviceSpecification )
+        {
+            result = paInvalidDevice;
+            goto error;
+        }
+
+        /* check that input device can support inputChannelCount */
+        if( inputChannelCount >
+            hostApi->deviceInfos[inputParameters->device]->maxInputChannels )
+        {
+            result = paInvalidChannelCount;
+            goto error;
+        }
+
+        /* validate inputStreamInfo */
+        if( inputParameters->hostApiSpecificStreamInfo )
+        {
+            /* this implementation doesn't use custom stream info */
+            result = paIncompatibleHostApiSpecificStreamInfo;
+            goto error;
+        }
+
+        hostInputSampleFormat =
+            PaUtil_SelectClosestAvailableFormat( inputSampleFormat,
+                                                 inputSampleFormat );
+
+        stream->inputFrameSize = Pa_GetSampleSize( inputSampleFormat ) * inputChannelCount;
+
+        result = PaPulseAudio_ConvertPortaudioFormatToPaPulseAudio_(
+            hostInputSampleFormat,
+            &stream->inSampleSpec );
+
+        if( result != paNoError )
+        {
+            goto error;
+        }
+
+        stream->inSampleSpec.rate = sampleRate;
+        stream->inSampleSpec.channels = inputChannelCount;
+        stream->latency = inputParameters->suggestedLatency;
+
+        if( !pa_sample_spec_valid(&stream->inSampleSpec) )
+        {
+            PA_DEBUG( ("Portaudio %s: Invalid input audio spec!\n",
+                      __FUNCTION__) );
+            result = paUnanticipatedHostError; /* should have been caught above */
+            goto error;
+        }
+
+        stream->inStream =
+            pa_stream_new( l_ptrPulseAudioHostApi->context,
+                           "Portaudio source",
```

Comment:
As one of the maintainers of the FreeDV project (which opens up to four sound devices at a time), it would be nice if we could override the "Portaudio source" and "Portaudio sink" descriptions so that people using pauvcontrol and similar can avoid confusion (for instance, [this comment](https://github.com/drowe67/freedv-gui/issues/16#issuecomment-817346121) while testing this PR with it). This seems like it could be done by providing a PulseAudio hostapi-specific API, much like what's done with e.g. ALSA.

That said, it's also a feature request that might be better suited for a future PR, especially as it's probably not too common to have more than one or two sound devices open at a time :) I'll play around a bit with it as well time-permitting and see if I can generate a patch that will accomplish this.

Reply:
For Mixxx we have the same use case.

Reply:
I'll take a look a that as it would be nice feature. My German is bit rusted but I'll get the point. Is this just cosmetic? As this is the case we should also provide API for http://0pointer.de/lennart/projects/pulseaudio/doxygen/proplist_8h.html but I think it's new PR case if someone wants it.

Reply:
I have some prototype code for stream renaming over at https://github.com/tmiw/portaudio/commit/d57f578a25a81d91afe650822fed6e1c7d154dc9 that seems to work okay with FreeDV. However, I imagine it'd need to be cleaned up and/or have more error checking before it can be included here. Hope that helps!

Reply:
@tmiw I have to eloborate little bit as this needs to be semipublic. @philburk can you check how to make this 'right' to have it on public header or something so applications can take advance of it?

Reply:
The mentioned commit has the changes to CMakeLists.txt, etc. to export the public include file for that new function. https://github.com/tmiw/portaudio/commit/b0577df6c79d05b6f5e9ba84523b7e2249fdbf04 has the new include file itself (I forgot to check it into the initial one before pushing). This was based on how the ALSA hostapi seemed to do it but I could definitely be missing stuff, so other feedback would be good.

Additionally, I could do a PR into your branch so that those changes can be included in this PR for review? Not sure if that'll work how I'm thinking.

Reply:
I think it would best to just support the standard PortAudio API first and then add the host specific APIs later.
Getting a basic stable release out early would be very beneficial. Then more people can start testing.
Also it will then become more clear, which, if any, host specific APIs are really needed. We try to minimize the host specific APIs. 

Also this PR is getting very big and could keep growing indefinitely. Let's try to wrap up this up with just the critical changes so we can merge it.

Reply:
I think the stream rename PR is close to being merged into this one, so we can probably include at least that. But yeah, the faster this one can be merged in the faster we can start getting runtime with other projects as well.

Reply:
@tmiw and @philburk I'm kind of middle of this. I would favor to get this in main tree sooner than later and these two left PR can be added then but then timeframe for them would be open. Otherwise there is at least two applications which wants to use this naming stuff. Biggest need for this PR would go away if there would be name for PaStream but that would be kind of big change to make (Besides that at least Jack would benefit from that) just for one.